static/js/services/endtoendencryption.js

```diff
@@ -0,0 +1,727 @@
+			} else {
+				// Need a way to store data for encryption support.
+				return;
+			}
+
+			this.axol = axolotl({
+				getLocalIdentityKeyPair: _.bind(function() {
+					return this.getLocalIdentityKeyPair.apply(this, arguments);
+				}, this),
+				getLocalRegistrationId: _.bind(function() {
+					return this.getLocalRegistrationId.apply(this, arguments);
+				}, this),
+				getLocalSignedPreKeyPair: _.bind(function() {
+					return this.getLocalSignedPreKeyPair.apply(this, arguments);
+				}, this),
+				getLocalPreKeyPair: _.bind(function() {
+					// Unsigned PreKeys are only used by protocol V2.
+					console.warn("Local PreKeys are not supported.")
+					return null;
+				}, this)
+			});
+			this.api.e.on("received.requestkeybundle",
+					_.bind(function(event, peer) {
+				this.requestKeyBundle(peer);
+			}, this))
+			this.api.e.on("received.keybundle",
+					_.bind(function(event, peer, data) {
+				this.receivedKeyBundle(peer, data);
+			}, this))
+			this.peer_identities = {};
+			this.sessions = {};
+			this.resetStore();
+		};
+
+		EndToEndEncryption.prototype.isSupported = function() {
+			return !!this.store;
+		};
+
+		EndToEndEncryption.prototype.resetStore = function() {
+			// TODO(fancycode): We should process any pending messages before
+			// resetting everything.
+			this.identity_keypair = null;
+			this.registration_id = null;
+			this.last_resort_pre_key = null;
+			this.next_signed_pre_key_id = 1;
+			this.pending_messages = {};
+			this.is_processing_messages = {};
+			this.ready_promise = null;
+			if (this.store.isReady()) {
+				_.defer(_.bind(function() {
+					this.register();
+				}, this));
+			}
+		};
+
+		EndToEndEncryption.prototype.setOwnIdentity = function(public_key) {
+			var identity = new PeerIdentity(null, public_key);
+			this.own_identity = identity;
+			this.events.triggerHandler("identity.own", [identity]);
+		};
+
+		EndToEndEncryption.prototype.storePeerIdentity = function(peer, public_key) {
+			var identity = new PeerIdentity(peer, public_key);
+			var fingerprint = identity.getFingerprint();
+			var existing = this.peer_identities[peer];
+			if (existing) {
+				if (existing.getFingerprint() === fingerprint) {
+					// No change.
+					return;
+				} else {
+					// Uh oh, remote peer has a new identity, this is something
+					// the user should know about!
+					this.events.triggerHandler("identity.changed", [
+						peer,
+						existing,
+						identity
+					]);
+				}
+			}
+			this.peer_identities[peer] = identity;
+			this.events.triggerHandler("identity.received", [peer, identity]);
+		};
+
+		EndToEndEncryption.prototype.getReadyPromise = function() {
+			if (!this.ready_promise) {
+				// Load/create initial objects
+				this.ready_promise = this.loadData().then(_.bind(function() {
+					var deferred = $q.defer();
+
+					_.defer(_.bind(function() {
+						var ready_state = READY_NONE;
+						var check_ready = _.bind(function() {
+							if (ready_state  === READY_ALL) {
```

Comment:
Guess it would be better to use `$q.all([this.getLocalIdentityKeyPair(), this.getLocalRegistrationId(), this.getLastResortPreKey()]).then(function() {` here instead, so we can get rid of the bitflags – Much cleaner imo.